site/source/_posts/2019-12-09-web3-what-are-your-options.md

```diff
@@ -0,0 +1,194 @@
+
+``` js
+var Web3 = require('web3');
+var web3 = new Web3('http://localhost:8545');
+```
+
+Something to note here, is that I'm calling `new Web3` with an `http` protocol, but the WebSocket protocol is also commonly used:
+
+``` js
+var web3 = new Web3(Web3.givenProvider || new Web3.providers.WebsocketProvider('ws://remotenode.com:8546'));
+``` 
+
+To test and ensure the connection, you can get a list of the wallets made available in Ganache by running:
+
+``` js
+web3.eth.getAccounts().then(console.log);
+```
+
+Which *should* give you an output like the following:
+
+```js
+> [ '0x7689cF9F90FAad61B8a3c9b1b2A5b4580B37358b',
+  '0x852e9a9db77a4e6169e175cDBb33dBE350150A8e',
+  '0x946700a1a4f30Dfe80307C91B6DaF1cCa2d09401',
+  '0x7d356aF02A87147D3ce5F9ACA716a78f70aF7616',
+  '0x88A116a16e4c8838F575a0e7a72eE27C7B073263',
+  '0x655317701Fcf3b310F492cB801C8D23f8c6fb556',
+  '0x16D305e72aFb0DDa1dB1830F8a98D5cD5337882E',
+  '0x9099bb4Af9CE5734E7a7e62f817e833fcFFaaF32',
+  '0x2ec4CC6700d0424A78a9B9Fc2ecBaeFc162313F1',
+  '0x1BC51a0edEC9FdEA3B14748e9209F4bF8Fe024b5' ]
+```
+
+If you want to check the balance of an individual wallet from the above list, you can do so by running:
+
+```js
+const account1 = "0x7689cF9F90FAad61B8a3c9b1b2A5b4580B37358b";
+
+web3.eth.getBalance(account1)
+.then(console.log);
+```
+
+Which will output:
+
+```
+> 100000000000000000000
+```
+
+
+## Contract Interaction
+
+As above; interacting with our *individual wallets* through `web3.js` is cool, but not nearly the extent to which the library works.  Let's now take a brief look at the more important functionality; of interacting with smart contracts through `web3.js`.
+
+The first thing we need to do, is to create a new contract, which we can do with the `new web3.eth.Contract` command.
+
+Before we call the `new` command, we need to assign our `json interface` for the contract's `ABI`:
+
+```js
+const abi = [{"type":"function", "name":"foo", "inputs": [{"name":"a","type":"uint256"}], "outputs": [{"name":"b","type":"address"}] },{ "type":"event", "name":"Event", "inputs": [{"name":"a","type":"uint256","indexed":true},{"name":"b","type":"bytes32","indexed":false}], }]
+```
+
+The `json interface` is a JSON object describing the *Application Binary Interface (ABI)* for our Ethereum smart contract.  Using this JSON interface; `web3.js` is able to create a JavaScript object representing our smart contract and its methods & events, using the `web3.eth.Contract` functionality.
+
+*Note, the above JSON interface / ABI is taken directly from the [Web3 docs](https://web3js.readthedocs.io/en/v1.2.0/web3-eth-contract.html#id5).*
+
+Now that we have our `json interface` defined, we can create our new contract instance: 
+
+```js
+var myContract = new web3.eth.Contract(abi, '0xde0B295669a9FD93d5F28D9Ec85E40f4cb697BAe');
+```
+
+*(The `from` address in this example is just random, but in a real case would be the address that all transactions of the contract were sent from.)*
```

Comment:
Maybe simpler: The address is the address of the already deployed contract instance that we're aiming to talk to